medicaltorch/datasets.py

```diff
@@ -214,26 +252,35 @@ def __init__(self, filename_pairs, slice_axis=2, cache=True,
         self._prepare_indexes()
 
     def _load_filenames(self):
-        for input_filename, gt_filename in self.filename_pairs:
-            segpair = SegmentationPair2D(input_filename, gt_filename,
-                                         self.cache, self.canonical)
-            self.handlers.append(segpair)
-
-    def _prepare_indexes(self):
-        for segpair in self.handlers:
-            input_data_shape, _ = segpair.get_pair_shapes()
-            for segpair_slice in range(input_data_shape[2]):
+        for input_filename, gt_filename, roi_filename, metadata, modality in self.filename_pairs:
+            segpair = SegmentationPair2D(input_filename, gt_filename, metadata=metadata,
+                                         cache=self.cache, canonical=self.canonical)
+            roipair = SegmentationPair2D(input_filename, roi_filename, metadata=metadata,
+                                         cache=self.cache, canonical=self.canonical)
 
-                # Check if slice pair should be used or not
-                if self.slice_filter_fn:
-                    slice_pair = segpair.get_pair_slice(segpair_slice,
-                                                        self.slice_axis)
+            self.handlers.append([segpair, roipair])
 
-                    filter_fn_ret = self.slice_filter_fn(slice_pair)
-                    if not filter_fn_ret:
+    def _prepare_indexes(self):
+        for seg_roi_pairs in self.handlers:
+            seg_pair, roi_pair = seg_roi_pairs
+            input_data_shape, _ = seg_pair.get_pair_shapes()
+
+            for idx_pair_slice in range(input_data_shape[self.slice_axis]):
+                slice_roi_pair = roi_pair.get_pair_slice(idx_pair_slice,
+                                                            self.slice_axis)
+                slice_seg_pair = seg_pair.get_pair_slice(idx_pair_slice,
+                                                            self.slice_axis)
+                # if ROI provided, filter empty (img, roi) slices
+                if self.slice_filter_fn and slice_roi_pair['gt'] is not None:
+                    filter_fn_ret_roi = self.slice_filter_fn(slice_roi_pair)
+                    if not filter_fn_ret_roi:
+                        continue
+                else:  # else filter empty (img, gt) slices
+                    filter_fn_ret_seg = self.slice_filter_fn(slice_seg_pair)
+                    if not filter_fn_ret_seg:
                         continue
 
-                item = (segpair, segpair_slice)
+                item = (segpair, roipair, segpair_slice)
```

Comment:
On line 283, (segpair, roipair, segpair_slice) are all undefined as well should it be (seg_pair, roi_pair, idx_pair_slice)

Reply:
Thanksfully you are here!

Reply:
Done